aea/protocols/default/serialization.py

```diff
@@ -17,53 +17,91 @@
 #
 # ------------------------------------------------------------------------------
 
-"""Serialization module for the default protocol."""
+"""Serialization module for default protocol."""
 
-import base64
-import json
 from typing import cast
 
 from aea.protocols.base import Message
 from aea.protocols.base import Serializer
+from aea.protocols.default import default_pb2
 from aea.protocols.default.message import DefaultMessage
 
 
 class DefaultSerializer(Serializer):
     """Serialization for the 'default' protocol."""
 
     def encode(self, msg: Message) -> bytes:
-        """Encode a 'default' message into bytes."""
+        """
+        Encode a 'Default' message into bytes.
+
+        :param msg: the message object.
+        :return: the bytes.
+        """
         msg = cast(DefaultMessage, msg)
-        body = {}  # Dict[str, Any]
-        body["type"] = msg.type.value
+        default_msg = default_pb2.DefaultMessage()
+        default_msg.message_id = msg.message_id
+        dialogue_reference = msg.dialogue_reference
+        default_msg.dialogue_starter_reference = dialogue_reference[0]
+        default_msg.dialogue_responder_reference = dialogue_reference[1]
+        default_msg.target = msg.target
 
-        if msg.type == DefaultMessage.Type.BYTES:
-            body["content"] = base64.b64encode(msg.content).decode("utf-8")
-        elif msg.type == DefaultMessage.Type.ERROR:
-            body["error_code"] = msg.error_code.value
-            body["error_msg"] = msg.error_msg
-            body["error_data"] = msg.error_data
+        performative_id = msg.performative
+        if performative_id == DefaultMessage.Performative.BYTES:
+            performative = default_pb2.DefaultMessage.Bytes()  # type: ignore
+            content = msg.content
+            performative.content = content
+            default_msg.bytes.CopyFrom(performative)
+        elif performative_id == DefaultMessage.Performative.ERROR:
+            performative = default_pb2.DefaultMessage.Error()  # type: ignore
+            error_code = msg.error_code
+            performative.error_code = error_code
```

Comment:
@5A11 this assignment does not actually work.

Reply:
Upon inspection, this line cannot be fixed quickly with the current setup (I didn't realise at first it is a custom type enum, so mistakenly told @DavidMinarsch I have a quick fix for it). 

In general, custom types are Not taken care of in the current generator. They appear in:
- `serialisation.py`
- `message.py`
- `X.proto`

They are an outstanding problem in `serialisation.py`. And even in the other two, we don't really properly tackle them.
- In `message.py`, we decided to take the class definitions out and put them in another module (`models.py` or `custom_types.py`) and have them `raise NotImplemented`.
- In `X.proto` we decided that their Protobuf definition is provided in the specification.

I think if a specification contains custom types we should either:
- generate the protocol package but warn the user and tell them to update the three files to cover the custom types, or 
- come up with a proper way to handle them (describing custom types in yaml/ontology) and generating the code in all the three files (`serialisation.py`, `message.py`, `X.proto`).